Lasso delayed feedbacks algorithm

jdematias opened the pull request:
Lasso delayed feedbacks algorithm

jdematias commented:
@VenturaDelMonte  
I know unit test is not completed but I would like that you give me your feedback about implementation

VenturaDelMonte commented on src/main/scala/eu/proteus/solma/lasso/LassoDFStreamTransformOperation.scala:
```diff
@@ -0,0 +1,104 @@
+    val labelStreamEvent = input.filter(x => x.isRight).map(x => x.right.get).head
+
+    val poses = inputStreamEvents.toList.map(x => x.pos._2)
+    var labels = List[(Double, Double)]()
```
im not sure a List is the cheapest data structure to use here.

VenturaDelMonte commented on src/main/scala/eu/proteus/solma/lasso/LassoDFStreamTransformOperation.scala:
```diff
@@ -0,0 +1,104 @@
+      val processedInput = rawInput.keyBy(x => selectKey(x)).window(GlobalWindows.create())
+        .trigger(new FitTrigger).apply(new FitWindowFunction())
+
+      val output = LassoParameterServer.transformLasso(None)(processedInput, workerParallelism = 3,
```
why do you bound the parallelism to 3? Wouldnt it be better to extract the parallelism from the execution environment?

VenturaDelMonte replied:
Moreover, there are a lot of hard coded params. This is ok at this stage, imho. However, we might need to rethink this part later on (or generalize those settings now).

VenturaDelMonte commented on src/main/scala/eu/proteus/solma/lasso/algorithm/FlatnessMappingAlgorithm.scala:
```diff
@@ -0,0 +1,43 @@
+  * TODO description
+  *
+  */
+class FlatnessMappingAlgorithm(input: List[Double],
```
I would suggest to check if the data structures used here are cheap enough w.r.t to the workload (e.g., index based lookups cost O(n) for a Scala List).

jdematias commented:
@VenturaDelMonte I have taken your suggestions into account and I have changed PR, but It's not totally finished. I have problem simulating real streaming working. I have two streams, one with features measurements and other with flatness measurements which are joined in a unique stream. The problem is that flatness should be processed always after all features measurements have been processed (like work in the real factory). Now I have added a "bad trigger" (trigger throw calculation when 70 events (a hardcoded value) are received) only to test the algorithm, but obviously this can not be the final solution. Could you suggest me a better way to simulate it?

VenturaDelMonte commented:
@jdematias I think the "custom trigger" you used to simulate is ok. An other solution would be to perform a join through a coMap/coFlatMap that would ingest both streams (say stream A containing "common" measurements and stream B containing flatness measurement). When items arrive to the operator  from stream A, they are immediately flushed downstream. Items arriving from stream B, they are cached. When measurements for a given coil are over, the cached flatness measurement for that coil is also sent downstream. We could track the ending of a coil through detecting a change of ids (we assume coils are produced in a FIFO fashion, right?). I believe this is a quick way to simulate our real scenario. What do you think about that?

jdematias commented:
Thanks @VenturaDelMonte !! It seems a good idea. I will try to code it and add new changes to PR

jdematias commented:
Hi @VenturaDelMonte
PR is OK for me. Please review it as soon as possible.
Thanks!!

VenturaDelMonte commented:
Great job! Merging and closing.
